### ISSUE TITLE
Race condition when closing device, causing core dump

### ISSUE THREAD

**Issue by dodgyrabbit** [abcminiuser/python-elgato-streamdeck#66]
**Describe the bug**
When calling close() there is a race condition that can cause and invalid handle to be accessed.

**To Reproduce**
Please see [this issue](https://github.com/timothycrosley/streamdeck-ui/issues/112#issuecomment-812922599) for details.

We are using version 0.8.3

The problem we saw is inside the **read** method of LibUSBHIDAPI.py (line 301)

``` python
def read(self, handle, length):
    if not handle:
        raise TransportError("No HID device.")

    data = ctypes.create_string_buffer(length)
            
    with self.mutex:
        result = self.hidapi.hid_read(handle, data, len(data))
```

The core problem is specifically that this line can run, with **handle** already invalid:
``` python
result = self.hidapi.hid_read(handle, data, len(data))
```

From the python trace output, we can see the following flow. I've colored the two threads blue (my application thread) and red (the key polling thread) to make it more obvious what I believe is going on:

![image](https://user-images.githubusercontent.com/518187/113493758-54099080-94b0-11eb-95c9-fa3a13473917.png)

**In summary:**

- The application is calling close() on the streamdeck.
- The handle is closed (line 195 blue) by the thread calling close(). Note the device_handle is not set to None **yet**.
- They _read loop calls into _read_key_states() which calls into read() (line 436 red).
- However, since self.device_handle is not set to None yet, it passes the stale handle as an argument to read().
- The check inside read() passes (it is not None). Since this is a parameter, it can't change even if the other thread sets it.
- The blue thread - now sets device_handle to None (doesn't really matter at this point, the red one can't observe it anyway).
- The red thread calls ` self.hidapi.hid_read(handle, data, len(data))` on a handle that is invalid.
- The read fails, exception is thrown.
Core dumped

It's not clear to me exactly what causes the core dump. Is it the read on an invalid handle, or is it the error handling or something.

**Comment by abcminiuser** [comment]
Thanks for the report!

I've moved the handle null checks into the lock in 69237993bffd63f025ed37c824b94304e3fa5a28 - please let me know if this helps, or if there's a second threading issue I need to look into.

**Comment by abcminiuser** [comment]
(Sorry, did this right before running out for the day - I see you edited your original message and there is indeed a deeper issue. I'll have to give this some thought and look at it tonight when I get back.

**Comment by dodgyrabbit** [comment]
Yes sorry on a second read of the code I realized that `handle` is a parameter and not a reference to the original value, hence moving the check into the mutex won't help :( 
Thanks for looking into this.

**Comment by abcminiuser** [comment]
Better fix in dcdc3e5941ed9d168db7914dc37999cc286dd17b. Since Python doesn't seem to have an atomic `swap()` equivalent I had to introduce a second mutex into the `Device` class, but overall the performance impact should be minimal.

**Comment by dodgyrabbit** [comment]
Thank you. I had a look and LGTM.

My understanding:
The mutex inside Library() protects against simultaneous access to the transport API. The mutex inside Device() protects against simultaneous read/writes of shared state, specifically, device_handle. Makes compare and set atomic.

I always worry about deadlocks with two or more locks, but provided we always acquire in the same order (Device->Library) we should be OK. Doesn't look like there's ever a callback while holding a lock either, so all good.

Will report back if it fixes the issue!

**Comment by abcminiuser** [comment]
Yes, that's correct. I don't believe there are deadlocks lurking here with the code as it stands now, as they're essentially nested locks (Device level locks first, then internally Library locks second). All I really needed was an atomic swap to atomically grab and `None`-out the device handle when closing, but alas, no such primitive seems to be available.

I've made this available as a 0.8.4 release for now - please do report back with positive or negative confirmation that the issue is solved.

**Comment by dodgyrabbit** [comment]
Great thanks, I've published a build of streamdeck-ui using 0.8.4 and gave it a quick runthrough. No obvious regressions or issues. Hopefully we'll know in the next day or so if it resolved the problem for the user who could produce the issue.
Your prompt response is greatly appreciated!

**Comment by dodgyrabbit** [comment]
Update:
The patch did not avoid the segfault :( 
It's not clear what causes the segfault so it may not be the streamdeck library after all.  However, I reviewed the latest trace and the fix you applied seems to work correctly. The handle is set to None and the read thread doesn't try operate on it.
